### PR TITLE
Integration tests for logins component

### PR DESCRIPTION
**Pull request by thomcc**
This was split off from #221. It's the start of a set of TPS-style tests. It has some issues but has been pretty valuable for me actually testing things.

### CODE REVIEW

**Comment by thomcc** [comment]
From meeting: Marked as in review, we should possibly do these discussions as followups.

**Comment by thomcc** [comment]
To elaborate on this:

It's basically a mini-TPS for the new sync stuff. Right now focusing on logins since that's the only new component that's finished. It's both simpler and better than TPS in a few ways:

1. The design of the new sync code as reusable libraries without global state allows us to easily instantiate two or more copies of the tests in the same process and sync them against eachother. No need for launching multiple processes and a driver program
2. It creates and destroys temporary accounts automatically, instead of relying on a cleanup process.

On the other hand, it doesn't use a clean high level/declarative description of the test phases and etc like TPS. They're straight imperative code (really TPS test files basically are too but it tries to hide this from you). This does allow for additional flexibility, but I'm not sure that actually matters (the more important thing is that it didn't require evaluating JavaScript from Rust).

It utilizes a helper program (`integration-test-helper/index.js`) to do so. This is a Node program that runs a headless browser to do the OAuth stuff, in theory all of it could be moved back into Rust if there were better browser automation libs for rust, but it looked like they were still rough.

Essentially the high level way it works is:

1. Create a temporary account. (`node integration-test-helper/index.js create <acct@restmail.net> <pass> <fxa_auth_url>`. This fetches from restmail and autoverifies if necessary)
2. Create some number of "clients" (something like a device or a session). For each client, log in with said temporary account over oauth (`node integration-test-helper/index.js oauth <acct@restmail.net> <pass> <oauth_url> <fxa_auth_url>` -- this launches a headless browser, and also does the restmail autoverify flow if necessary)
3. For each test case:
    - Run the test case
    - Clear local state
    - Wipe the server afterwards
4. Delete the temporary account at the end (`node integration-test-helper/index.js destroy <acct@restmail.net> <pass> <fxa_auth_url>`).

At the moment it runs as a rust integration test (e.g. a rs file under `tests`), I think we probably will want to move it into a separate program that we run as part of CI, so that we don't end up making `cargo test --all` very slow. It's also at the point where Cargo emits a warning about it being a "slow test". This is partially because the create/login flow is slow, but also because we run all tests using the same temporary account, which means it needs to be a single test function.

It looks like rust's upcoming support for [custom test frameworks](https://github.com/rust-lang/rfcs/blob/master/text/2318-custom-test-frameworks.md) (see also [this description](https://blog.jrenner.net/rust/testing/2018/08/06/custom-test-framework-prop.html) which makes it more clear to me how this would help a tps-like tool) would make something like this doable in a much more clean and high level way. In the short term, we can't really justify waiting for that to come into being, so this is neither here nor there.

@davehunt, @ncalexan suggested we should get your feedback before moving forward with this. Not sure how technical you would want to get on it, but certainly if you want to review the actual code, by all means.

File: logins-sql/tests/helpers/mod.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    pub fn new_random() -> FailureResult<Arc<TestAccount>> {
+        use rand::{self, prelude::*};
+        let mut rng = thread_rng();
+        let name = format!("rust-login-sql-test--{}",
+            rng.sample_iter(&rand::distributions::Alphanumeric).take(5).collect::<String>());
+        // Just use the username for the password in case we need to clean these
+        // up easily later because of some issue.
+        let password = name.clone();
+        let email = format!("{}@restmail.net", name);
+        Self::new(email, password, FxaConfig::stable_dev()?)
```
Comment (vladikoff) [inline logins-sql/tests/helpers/mod.rs]: `stable` cannot pass the tests at this time for strange reasons, but `latest` should work, if we want to land this then we should change this for now

**Comment by thomcc** [comment]
@ncalexan This is now a separate executable, as requested. You mentioned before that you'd like some sort of CLI options, but I'm not sure what it should take, so I've left it out for now. In the future as there are more tests adding options likely will make more sense.

File: sync-test/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+#[macro_use]
+extern crate lazy_static;
+
+mod support;
+
+use support::*;
+
+use logins_sql::Login;
+
+macro_rules! cleanup_clients {
```
Comment (ncalexan) [inline sync-test/src/main.rs]: This seems fancy for what it's actually doing, but whatever.

File: sync-test/src/support.rs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    // This `Once` is used to run `npm install` first time through.
+    static HELPER_SETUP: Once = ONCE_INIT;
+    HELPER_SETUP.call_once(|| {
+        let dir = &*HELPER_SCRIPT_DIR;
+        env::set_current_dir(dir).expect("Failed to change directory...");
+
+        // Let users know why this is happening even if `log` isn't enabled.
+        println!("Running `npm install` in `integration-test-helper` to ensure it's usable");
+
+        let mut child = Command::new("npm")
```
Comment (ncalexan) [inline sync-test/src/support.rs]: Can we take `npm` and `node` from `$NPM` and `$NODE`, falling back to `"npm"` and `"node"`?  (Let's ease our automation configuration up front.)

**Comment by thomcc** [comment]
@eoger can you take a look at the circle-ci changes? (e.g. 793002c)

**Comment by thomcc** [comment]
@eoger Can I have permission to retrigger jobs on circleci? (I guess I don't have this)

**Comment by thomcc** [comment]
That said, this passes locally reliably, and I don't want to introduce an intermittent failure so probably some debugging is needed before landing (damn)

**Comment by thomcc** [comment]
This seems to have stopped failing so I'm just going to land it and hope for the best.

If it starts failing again we can do something about that.

**Pull request merged**